File: go-backend-interview-prep/01_go_basics/12_packages_and_imports/main.go
package main

import (
	"fmt"
	"math"
	"strings"
)

func main() {
	// --- Using the fmt package ---
	fmt.Println("--- fmt package ---")
	fmt.Println("Println adds a newline")
	fmt.Printf("Printf with formatting: %d\n", 42)

	// --- Using the strings package ---
	fmt.Println("\n--- strings package ---")
	msg := "Hello, Go World"

	// Exported functions start with uppercase
	fmt.Println("ToUpper:", strings.ToUpper(msg))
	fmt.Println("ToLower:", strings.ToLower(msg))
	fmt.Println("Contains 'Go':", strings.Contains(msg, "Go"))
	fmt.Println("Replace:", strings.ReplaceAll(msg, "Go", "Gopher"))
	fmt.Println("Split:", strings.Split(msg, " "))

	// --- Using the math package ---
	fmt.Println("\n--- math package ---")
	fmt.Println("Pi:", math.Pi)     // Exported constant (uppercase P)
	fmt.Println("Sqrt(16):", math.Sqrt(16))
	fmt.Println("Max(3,7):", math.Max(3, 7))

	// --- Visibility rule ---
	fmt.Println("\n--- Visibility rule ---")
	fmt.Println("math.Pi is exported (uppercase P)")
	// math.pi would NOT compile -- lowercase = unexported

	// --- Package name = last segment ---
	// "math" package -> math.Pi
	// "strings" package -> strings.ToUpper
	// "fmt" package -> fmt.Println
	fmt.Println("\nPackage names match the last import path segment.")

	// --- Internal functions demonstrate visibility ---
	fmt.Println("\n--- Local visibility ---")
	fmt.Println("exported call:", ExportedGreet("Alice"))
	fmt.Println("unexported call:", unexportedHelper())
}

// ExportedGreet is visible outside this package (uppercase E)
func ExportedGreet(name string) string {
	return "Hello, " + name
}

// unexportedHelper is only visible within this package (lowercase u)
func unexportedHelper() string {
	return "I am package-private"
}

File: go-backend-interview-prep/01_go_basics/12_packages_and_imports/main.py
# Python equivalent of Packages and Imports (compare with main.go)
# Python uses modules and packages (directories with __init__.py).

import math
import string

def main():
    # --- Using the math module ---
    print("--- math module ---")
    print("pi:", math.pi)
    print("sqrt(16):", math.sqrt(16))

    # --- Using the string module ---
    print("\n--- string module ---")
    msg = "Hello, Python World"
    print("upper:", msg.upper())       # method on str, not a module function
    print("lower:", msg.lower())
    print("contains 'Python':", "Python" in msg)
    print("replace:", msg.replace("Python", "Pythonista"))
    print("split:", msg.split(" "))

    # --- Visibility rules ---
    # Python convention: _prefix means private, no prefix means public.
    # Go: uppercase = exported, lowercase = unexported.
    print("\n--- Visibility ---")
    print("public function:", exported_greet("Alice"))
    print("private function:", _unexported_helper())

    # --- Module name = file name ---
    # Go: package name = last segment of import path
    # Python: module name = file name (without .py)
    print("\nPython module names = file names (not path segments).")


# Public function (no underscore prefix)
def exported_greet(name: str) -> str:
    return "Hello, " + name


# Private function (underscore prefix, convention only)
def _unexported_helper() -> str:
    return "I am module-private (by convention)"


if __name__ == "__main__":
    main()
